File: 1189-maximum-number-of-balloons/solution.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>

class Solution {
public:
    int maxNumberOfBalloons(std::string text) {
        // Frequency map to store counts of characters in text
        std::vector<int> counts(26, 0);
        for (char c : text) {
            counts[c - 'a']++;
        }

        // We need: b:1, a:1, l:2, o:2, n:1
        // We calculate how many full sets of each we have
        int b = counts['b' - 'a'];
        int a = counts['a' - 'a'];
        int l = counts['l' - 'a'] / 2;
        int o = counts['o' - 'a'] / 2;
        int n = counts['n' - 'a'];

        // The result is the bottleneck (the minimum of these requirements)
        return std::min({b, a, l, o, n});
    }
};

File: 1189-maximum-number-of-balloons/solution.py
from collections import Counter

class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        # Count frequency of each character in the input text
        counts = Counter(text)
        
        # Calculate how many of each required character we have
        # For 'l' and 'o', we need 2 of each per word
        b = counts['b']
        a = counts['a']
        l = counts['l'] // 2
        o = counts['o'] // 2
        n = counts['n']
        
        # The maximum number of balloons is limited by the smallest frequency
        return min(b, a, l, o, n)